### PR TITLE
Initialize a Gradle plugin project with Kotlin

### LINKED ISSUE #1: Initialize a Gradle plugin project with Kotlin
Initialize a Gradle plugin project. It should use Kotlin as it's language.

### PR DESCRIPTION
**Pull request by mortenjuult**
Initial commit of the gradle plugin

Resolves #1

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+group 'no.item.gradle'
+version '1.0.0'
+
+sourceCompatibility = 1.8
+targetCompatibility = 1.8
+
+repositories {
+    mavenCentral()
+    jcenter()
+   // maven(url = "https://dl.bintray.com/arrow-kt/arrow-kt/")
```
Comment (tajakobsen) [inline build.gradle]: Don't check in commented out code, just remove it, or comment it in again

File: src/main/kotlin/no/item/gradle/InitRunnerPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package no.item.gradle
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+
+class InitRunnerPlugin : Plugin<Project> {
+    override fun apply(target: Project) {
+        val extension = target.extensions.create(
+            "initRunner",
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/InitRunnerPlugin.kt]: Lets rename `"initRunner"` to `"generateTypeScript"`.

File: src/main/kotlin/no/item/gradle/InitRunnerTask.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package no.item.gradle
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+import java.io.File
+import javax.inject.Inject
+
+open class InitRunnerTask @Inject constructor(private val extension: InitRunnerExtension) : DefaultTask(){
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/InitRunnerTask.kt]: If this task is listing xml-files, maybe we should rename the file (and class)  to `FindXmlFilesTask.kt`?

File: src/main/kotlin/no/item/gradle/InitRunnerTask.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    @TaskAction
+    fun findXmlFiles() {
+        val list: MutableList<String> = ArrayList()
+        File(extension.fileDir).walk().forEach {
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/InitRunnerTask.kt]: I think using a `filter` and `map` would be nicer then mutating a list. It indicates intent more clearly then a `forEach`.

As far as possible, lets try to minimize the amount of mutation we do. Keep closer to a functional programming paradigm.

```kotlin
val files = File(extension.fileDir)
    .walk()
    .filter { it.isFile && it.extension == "xml" }
    .map { it.absolutePath }
```

File: src/main/kotlin/no/item/gradle/InitRunnerTask.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+            if (it.isFile&&it.toString().endsWith(".xml")){
+                list.add(it.absolutePath);
+            }
+        }
+        println("Found XML files: ")
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/InitRunnerTask.kt]: We can keep this for now. But later I suggest we only do `println` to indicate that we generate a new file, or that we update an existing file.

File: src/main/kotlin/no/item/gradle/InitRunnerExtension.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import javax.inject.Inject
+
+@Suppress("UnstableApiUsage") // ObjectFactory is incubating
+open class InitRunnerExtension @Inject constructor(
+    objectFactory: ObjectFactory
+) {
+
+    // region Model
+
+    var coverageTypes = listOf("instruction")
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/InitRunnerExtension.kt]: Just remove this, if it's not intended to be used for what we are doing.

Comment (tajakobsen) [reply src/main/kotlin/no/item/gradle/InitRunnerExtension.kt]: Configs we might be interested in taking in, includes a black list over xml-files we don't want to generate code from.

File: src/main/kotlin/no/item/gradle/InitRunnerPlugin.kt (hunk 2 above)
Comment (tajakobsen) [reply src/main/kotlin/no/item/gradle/InitRunnerPlugin.kt]: And also rename the files themselves.

**Review by tajakobsen — CHANGES_REQUESTED** [review]
Good job!

Here are a few things to look at:
 - Don't check in the `.idea`-directory (add it to the .gitignore file)
 - Add information about how to build, and how to use the plugin in README.md

File: src/main/kotlin/no/item/gradle/beans/InputType.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package no.item.gradle.beans
+
+enum class InputType {
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/beans/InputType.kt]: The convention for `enum` is that constants are upper case. Ref. https://kotlinlang.org/docs/reference/enum-classes.html

File: src/main/kotlin/no/item/gradle/beans/GeneratedField.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package no.item.gradle.beans
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/beans/GeneratedField.kt]: Are `data classes` really `beans`? I would possibly name this package `models`.

  - no.item.xp.plugin
  - no.item.xp.plugin.models
  - no.item.xp.plugin.parsing
  - no.item.xp.plugin.writer

What do you think?

File: demo/app-report.xml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (tajakobsen) [inline demo/app-report.xml]: Move demo folder out of project.

File: build.gradle [hunk 10]
```diff
@@ -0,0 +1,53 @@
+plugins {
+    id 'java'
+    id 'org.jetbrains.kotlin.jvm' version '1.3.61'
+    id 'java-gradle-plugin'
+    id 'maven'
+}
+
+group 'no.item.gradle'
```
Comment (tajakobsen) [inline build.gradle]:
```suggestion
group 'no.item.xp.plugin'
```

File: src/main/kotlin/no/item/gradle/GenerateTypeScriptTask.kt [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      println("Found XML files: ")
+      println("--------------------")
+      files.forEach { println(it) }
+      //2.send dem til en parser
+      //3.send resultatet til en writer
+      //4.exit
+    }
+
+  fun findXmlFiles(): Sequence<String> {
+    val files = File(extension.fileDir)
```
Comment (tajakobsen) [inline src/main/kotlin/no/item/gradle/GenerateTypeScriptTask.kt]: I would return this directly without setting it to `val files` first.

**Review by tajakobsen — CHANGES_REQUESTED** [review]
Good job! We are getting closer. I don't see the Kotlin-linter I send you a message about. It would be great to just be able to to some linting (and automatical fixing) on the code before submitting it.

File: src/main/kotlin/no/item/gradle/InitRunnerPlugin.kt (hunk 2 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/InitRunnerPlugin.kt]: Good idea, implemented

File: src/main/kotlin/no/item/gradle/InitRunnerTask.kt (hunk 3 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/InitRunnerTask.kt]: Good idea, implemented

File: src/main/kotlin/no/item/gradle/InitRunnerTask.kt (hunk 4 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/InitRunnerTask.kt]: Good idea, implemented

File: src/main/kotlin/no/item/gradle/InitRunnerExtension.kt (hunk 6 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/InitRunnerExtension.kt]: Removed a bunch of unnecessary code here. For now the class is moslty empty, but keeping it for possibilty of configs later.

File: src/main/kotlin/no/item/gradle/InitRunnerTask.kt (hunk 5 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/InitRunnerTask.kt]: Yeah, this kind of printing will be removed once the code is out of infancy.

File: src/main/kotlin/no/item/gradle/beans/InputType.kt (hunk 7 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/beans/InputType.kt]: Alright, lets keep it to convention.

File: src/main/kotlin/no/item/gradle/beans/GeneratedField.kt (hunk 8 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/beans/GeneratedField.kt]: No they are not beans. Models are a better name,  will be changed.

File: demo/app-report.xml (hunk 9 above)
Comment (mortenjuult) [reply demo/app-report.xml]: demo-folder deleted

File: build.gradle (hunk 10 above)
Comment (mortenjuult) [reply build.gradle]: Yeah, i like this. Renamed now.

File: src/main/kotlin/no/item/gradle/GenerateTypeScriptTask.kt (hunk 11 above)
Comment (mortenjuult) [reply src/main/kotlin/no/item/gradle/GenerateTypeScriptTask.kt]: I agree.

**Pull request merged**